exporter/prometheusexporter/metric_rename.go

```diff
@@ -0,0 +1,198 @@
+// Prometheus best practices for units: https://prometheus.io/docs/practices/naming/#base-units
+// OpenMetrics specification for units: https://github.com/OpenObservability/OpenMetrics/blob/main/specification/OpenMetrics.md#units-and-base-units
+var unitMap = map[string]string{
+
+	// Time
+	"d":   "days",
+	"h":   "hours",
+	"min": "minutes",
+	"s":   "seconds",
+	"ms":  "milliseconds",
+	"us":  "microseconds",
+	"ns":  "nanoseconds",
+
+	// Bytes
+	"By":   "bytes",
+	"KiBy": "kibibytes",
+	"MiBy": "mebibytes",
+	"GiBy": "gibibytes",
+	"TiBy": "tibibytes",
+	"KBy":  "kilobytes",
+	"MBy":  "megabytes",
+	"GBy":  "gigabytes",
+	"TBy":  "terabytes",
+	"B":    "bytes",
+	"KB":   "kilobytes",
+	"MB":   "megabytes",
+	"GB":   "gigabytes",
+	"TB":   "terabytes",
+
+	// SI
+	"m": "meters",
+	"V": "volts",
+	"A": "amperes",
+	"J": "joules",
+	"W": "watts",
+	"g": "grams",
+
+	// Misc
+	"Cel": "celsius",
+	"Hz":  "hertz",
+	"1":   "",
+	"%":   "percent",
+	"$":   "dollars",
+}
+
+// The map that translates the "per" unit
+// Example: s => per second (singular)
+var perUnitMap = map[string]string{
+	"s":  "second",
+	"m":  "minute",
+	"h":  "hour",
+	"d":  "day",
+	"w":  "week",
+	"mo": "month",
+	"y":  "year",
+}
+
+// Build a Prometheus-compliant metric name for the specified metric
+func buildPromCompliantName(metric pmetric.Metric) string {
+
+	// Split metric name in "elements" (remove all non-alphanumeric)
+	nameElements := strings.FieldsFunc(
+		metric.Name(),
+		func(r rune) bool { return !unicode.IsLetter(r) && !unicode.IsDigit(r) },
+	)
+
+	// Clean-up the metric's unit
+	// Remove all {xyz} elements from unit
+	cleanUnit := removeBracketedUnits(strings.TrimSpace(metric.Unit()))
+
+	// Split at the '/' if any
+	unitElements := strings.SplitN(cleanUnit, "/", 2)
+
+	// Main unit
+	if len(unitElements) > 0 && unitElements[0] != "" {
+		mainUnit := cleanUpString(unitMapGetOrDefault(unitElements[0]))
+		if mainUnit != "" && !contains(nameElements, mainUnit) {
+			nameElements = append(nameElements, mainUnit)
+		}
+	}
+
+	// Per unit
+	if len(unitElements) > 1 && unitElements[1] != "" {
+		perUnit := cleanUpString(perUnitMapGetOrDefault(unitElements[1]))
+		if perUnit != "" && !contains(nameElements, perUnit) {
+			nameElements = append(append(nameElements, "per"), perUnit)
+		}
+	}
+
+	// Append _total for Counters
+	if metric.DataType() == pmetric.MetricDataTypeSum && metric.Sum().IsMonotonic() {
+		nameElements = append(removeItem(nameElements, "total"), "total")
+	}
+
+	// Append _ratio for ratio Gauges
+	if cleanUnit == "1" && metric.DataType() == pmetric.MetricDataTypeGauge {
+		nameElements = append(removeItem(nameElements, "ratio"), "ratio")
+	}
+
+	return strings.Join(nameElements, "_")
+}
+
+// Remove all {xyz} elements from specified string
+func removeBracketedUnits(unit string) string {
```

Comment:
We expect `{xyz}` to be the entire word, right?  E.g. `{bananas}/s`.  Could we just check to see if any part of `unitElements` contains `{` or `}`, and map it to empty-string if that is the case?

Reply:
I guess we can simplify here. Thanks!